compiler/lquantizer/src/LQBinarizer.cpp

```diff
@@ -0,0 +1,264 @@
+    for (int32_t b = 0; b < encode_bits; ++b)
+    {
+      int32_t bit_value = (data_binary[b * data_bin_size + idx] >> offset) & 1;
+      dequantized_data[i] += (bit_value == 1) ? scales[b] : -scales[b];
+    }
+  }
+
+  return dequantized_data;
+}
+
+void LQBinarizer::dequantize(const std::vector<int8_t> &bins, std::vector<float> &data)
+{
+  // calculate data from binary
+  for (int32_t i = 0; i < data_float_size; ++i)
+  {
+    data[i] = 0;
+    for (int32_t b = 0; b < encode_bits; ++b)
+    {
+      data[i] += static_cast<float>(bins[i * encode_bits + b]) * scales[b];
+    }
+  }
+}
+
+// Method returns bits in float representation for data value with index i
+std::vector<float> LQBinarizer::value_bits_as_floats(int32_t i)
+{
+  std::vector<float> bits(encode_bits);
+
+  int32_t idx = i >> 5;    // divide to 32
+  int32_t offset = i & 31; // mod of 32
+
+  for (int32_t b = 0; b < encode_bits; ++b)
+  {
+    int32_t bit_value = (data_binary[b * data_bin_size + idx] >> offset) & 1;
+    bits[b] = (bit_value == 1) ? 1.0f : -1.0f;
+  }
+
+  return bits;
+}
+
+void LQBinarizer::unpack_binary(std::vector<int8_t> &buffer)
+{
+  assert(buffer.size() == encode_bits * data_float_size);
+  for (int32_t i = 0; i < data_float_size; ++i)
+  {
+    int8_t *bits = &buffer[i * encode_bits];
+
+    int32_t idx = i >> 5;    // divide to 32
+    int32_t offset = i & 31; // mod of 32
+
+    for (int32_t b = 0; b < encode_bits; ++b)
+    {
+      int32_t bit_value = (data_binary[b * data_bin_size + idx] >> offset) & 1;
+      bits[b] = (bit_value == 1) ? 1 : -1;
+    }
+  }
+}
+
+// iterative gradient descent to search optimal scales `v: argmin||Bxv - x||^2`, `x` - data_vector
+#if 0
+void LQBinarizer::gradient_descent_scales(const float *data_vector)
+{
+  for (int32_t step = 0; step < descent_steps; ++step)
+  {
+    std::vector<float> Bxv = dequantize();
+    std::vector<float> dv(encode_bits, 0.0f);
+
+    // search dv
+    for (uint32_t i = 0; i < Bxv.size(); ++i)
+    {
+      std::vector<float> grad = value_bits_as_floats(i);
+
+      float err = Bxv[i] - data_vector[i];
+      float coef = descent_lr / Bxv.size();
+
+      for (int32_t b = 0; b < encode_bits; ++b)
+      {
+        grad[b] *= err;
+        dv[b] -= coef * grad[b]; // add antigradient with coef to dv
+      }
+    }
+
+    // correct scales with dv
+    for (int32_t b = 0; b < encode_bits; ++b)
+    {
+      scales[b] += dv[b];
+    }
+  }
+}
+#else
+// iterative gradient descent to search optimal scales `v: argmin||Bxv - x||^2`, `x` - data_vector
+void LQBinarizer::gradient_descent_scales(const float *data_vector)
+{
+  // transposed binary vector with single char for bit
+  std::vector<int8_t> bin_trans(encode_bits * data_float_size);
+  unpack_binary(bin_trans);
+
+  // allocate buffer for dequantized values and gradient step
+  std::vector<float> bin2fp(data_float_size);
+  std::vector<float> dv(encode_bits);
+
+  float const coef = descent_lr / static_cast<float>(data_float_size);
```

Comment:
current bug - somitimes this value is nan due to `data_float_size` equal to 0. TODO resolve this bug in `input_train() ` methid of Quantizer.